legacy/reqcounter/reqcounter.go

```diff
@@ -16,30 +16,73 @@ limitations under the License.
 
 package reqcounter
 
-import "sync"
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
 
 // RequestCounter records the number of HTTP requests to GCR.
 type RequestCounter struct {
-	mutex    sync.Mutex
-	requests uint64
+	mutex    sync.Mutex // Lock to prevent race-conditions with concurrent processes.
+	requests uint64     // Number of HTTP requests since recording started.
+	quota    uint64     // Maximum requests allowed before GCR throttling occurs.
+	since    time.Time  // When the current counter began recording requests.
+}
+
+// log records the number of HTTP requests found and resets the counter.
+func (rc *RequestCounter) log() {
+	// Grab the lock to prevent race-conditions.
+	rc.mutex.Lock()
+	defer rc.mutex.Unlock()
+	// Log the number of requests found.
+	msg := fmt.Sprintf("Since %s there have been %d requests to GCR.", counter.since.Format("01-02-2006 15:04:05"), counter.requests)
+	logrus.Debug(msg)
+	// Reset the counter.
+	rc.reset()
+}
+
+// reset clears the number of requests and stamps the object with the current time of reset.
+// WARNING: To avoid race conditions, only use this method when holding the counter's mutex.
```

Comment:
It's not safe to release the lock and expect to immediately grab it again. Here is an example of what could happen:

0. log grabs the lock
1. log collects the requests
2. log lets go of the lock
3. Increment grabs the lock
4. Increment adds 1 to the already logged counter
5. Increment releases the lock
6. reset grabs the lock
7. reset wipes the counter
8. reset lets go of the lock

In this scenario, we reset the lock with an extra request that was not logged. To avoid this, the logic must remain how it is, or we add the `reset` logic inside of the `log` function.

Reply:
Maybe this is more complicated than it needs to be? What if we only lock during writes to the counter, not reads?

That is, we could remove step 0 and 2 in your example (i.e., make `log` read the requests counter without using a lock); this loses accuracy down to 1 (maybe 2?) requests, but I think it's a fine because (1) this off-by-1 or 2 error does not accumulate over time (that is, it will never become an off-by-1000 or 2000 error), and (2) we don't even know when the GCP reference 10-minute interval starts/stops (even if we did, we won't be able to synchronize our clock to match GCP's exactly anyway), so our own 10-minute interval is a best-guess effort.

Reply:
This still doesn't seem to be the way to handle race-conditions. The simplest approach, that still feels most correct, is to just combine the reset logic and let go of the lock when reading and writing has ended. Race conditions are eliminated and there is one less function to handle.